False positive in CA1307 for string.IndexOf(char)

#### Analyzer package

Microsoft.CodeAnalysis.FxCopAnalyzers

#### Package Version

v2.9.3

#### Diagnostic ID

[CA1307](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1307-specify-stringcomparison): Specify StringComparison

#### Repro steps

Compile the following code against .NET Core 2.2:

    class C
    {
        int M(string s) => s.IndexOf('?');
    }

#### Expected behavior

No warning.

#### Actual behavior

Warning: The behavior of 'string.IndexOf(char)' could vary based on the current user's locale settings. Replace this call in 'C.M(string)' with a call to 'string.IndexOf(char, System.StringComparison)'.

The premisse of the diagnostic, that string.indexOf(char) could vary is incorrect. The [method documentation](https://docs.microsoft.com/en-us/dotnet/api/system.string.indexof?view=netcore-2.2#System_String_IndexOf_System_Char_) clearly states:

> This method performs an ordinal (culture-insensitive) search, where a character is considered equivalent to another character only if their Unicode scalar values are the same.

Note also that replacing the `s.IndexOf('?')` by `s.IndexOf('?', StringComparison.Ordinal)` is undesirable, as the method is less efficient.

Here's my proposal for how to modify the rule to increase the signal to noise ratio.

For the methods `string.StartsWith`, `string.EndsWith`, `string.IndexOf`, and `string.LastIndexOf` specifically:

 * If the first parameter (after _this_) to the method is a `string`, and if the target overload doesn't accept a `CultureInfo` or a `StringComparison` parameter, produce warning CA1307 and tell the developer that they should call an overload that takes a `CultureInfo` or a `StringComparison`.

 * If the first parameter (after _this_) to the method is a `char` (or anything other than `string`), or if they're calling an overload which accepts a `CultureInfo` or a `StringComparison` parameter, then CA1307 should not trigger.

For the methods `string.ToUpper` and `string.ToLower` specifically:

 * If calling the parameterless overload, produce warning CA1307 and tell the developer that they should instead call `ToUpperInvariant` or `ToLowerInvariant` or that they should call a `ToUpper` or `ToLower` overload which takes a `CultureInfo`.

For the method `string.Compare` specifically:

 * If calling an overload that does not take a `CultureInfo` or `StringComparison` parameter, produce warning CA1307 and tell the developer that they should call an overload which accepts a `CultureInfo` or `StringComparison` parameter.

 * If calling an overload that takes a `CultureInfo` or `StringComparison` parameter, warning CA1307 should not trigger.

For the method `string.CompareTo` specifically:

 * Produce warning CA1307 and tell the developer to call `string.Compare` instead, calling an overload which accepts a `CultureInfo` or `StringComparison` parameter.

No method on `string` other than those methods specifically called out above should produce warning CA1307.

The intent of this proposal is to identify places in the `string` API surface where the current culture is implicitly captured and used as part of the operation. It is not always intuitive for the developer to know which `string` APIs are culture-sensitive by default and which APIs are ordinal by default. These warnings also make it easier to port existing `string`-based code to `ReadOnlySpan<char>`-based patterns, as all of the span-based APIs operate under an ordinal comparison by default unless an explicit comparison has been passed to the API.

--- a/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyStringComparison.cs
+++ b/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/SpecifyStringComparison.cs
@@ -102,6 +102,7 @@ namespace Microsoft.NetCore.Analyzers.Runtime
                         return;
                     }
 
+                    // First check if this is known method with a recommended alternate overload that must be flagged.
                     if (!overloadMap.IsEmpty && overloadMap.ContainsKey(targetMethod))
                     {
                         ReportDiagnostic(
@@ -111,6 +112,21 @@ namespace Microsoft.NetCore.Analyzers.Runtime
                             overloadMap[targetMethod]);
 
                         return;
+                    }
+
+                    // Otherwise, we want to generically flag any method that has an additional overload with the exact same parameter list,
+                    // plus as additional StringComparison parameter at the end.
+
+                    if (targetMethod.ContainingType.SpecialType == SpecialType.System_String)
+                    {
+                        // We do not want to generically report for methods on 'string' type where the first parameter is not a 'string' type.
+                        // For non-string related comparisons, this can lead to the rule being very noisy as such overloads have 
+                        // StringComparison.Ordinal as the default, and that is desirable majority of times.
+                        // See https://github.com/dotnet/roslyn-analyzers/issues/2581 for details
+                        if (targetMethod.Parameters.IsEmpty || targetMethod.Parameters[0].Type.SpecialType != SpecialType.System_String)
+                        {
+                            return;
+                        }
                     }
 
                     IEnumerable<IMethodSymbol> methodsWithSameNameAsTargetMethod = targetMethod.ContainingType.GetMembers(targetMethod.Name).OfType<IMethodSymbol>();

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SpecifyStringComparisonTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SpecifyStringComparisonTests.cs
@@ -2,6 +2,7 @@
 
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis.Testing;
+using Test.Utilities;
 using Xunit;
 using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
     Microsoft.NetCore.Analyzers.Runtime.SpecifyStringComparisonAnalyzer,
@@ -55,7 +56,7 @@ GetCA1307CSharpResultsAt(14, 18, $"string.Compare({StringArgType}, int, {StringA
         }
 
         [Fact]
-        public async Task CA1307_StringWithTests_CSharp()
+        public async Task CA1307_StringWithStringTests_CSharp()
         {
             await VerifyCS.VerifyAnalyzerAsync(@"
 using System;
@@ -79,8 +80,25 @@ GetCA1307CSharpResultsAt(12, 16, "string.StartsWith(string)",
                                  "string.StartsWith(string, System.StringComparison)"));
         }
 
+        [Fact, WorkItem(2581, "https://github.com/dotnet/roslyn-analyzers/issues/2581")]
+        public async Task CA1307_StringWithCharTests_CSharp()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+using System;
+using System.Globalization;
+
+public class StringComparisonTests
+{
+    public bool StringWith(string strA, char chA, char chB)
+    {
+        var x = strA.EndsWith(chA);
+        return strA.StartsWith(chB);
+    }
+}");
+        }
+
         [Fact]
-        public async Task CA1307_StringIndexOfTests_CSharp()
+        public async Task CA1307_StringIndexOfStringTests_CSharp()
         {
             await VerifyCS.VerifyAnalyzerAsync(@"
 using System;
@@ -105,6 +123,40 @@ GetCA1307CSharpResultsAt(11, 18, "string.IndexOf(string, int)",
 GetCA1307CSharpResultsAt(12, 16, "string.IndexOf(string, int, int)",
                                  "StringComparisonTests.StringIndexOf()",
                                  "string.IndexOf(string, int, int, System.StringComparison)"));
+        }
+
+        [Fact, WorkItem(2581, "https://github.com/dotnet/roslyn-analyzers/issues/2581")]
+        public async Task CA1307_StringIndexOfCharTests_CSharp()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+using System;
+using System.Globalization;
+
+public class StringComparisonTests
+{
+    public int StringIndexOf(string strA, char chA)
+    {
+        var x1 = strA.IndexOf(chA);
+        var x2 = strA.IndexOf(chA, 0);
+        return strA.IndexOf(chA, 0, 1);
+    }
+}");
+        }
+
+        [Fact, WorkItem(2581, "https://github.com/dotnet/roslyn-analyzers/issues/2581")]
+        public async Task CA1307_StringGetHashCodeTests_CSharp()
+        {
+            await VerifyCS.VerifyAnalyzerAsync(@"
+using System;
+using System.Globalization;
+
+public class StringGetHashCodeTests
+{
+    public int StringGetHashCode(string strA)
+    {
+        return strA.GetHashCode();
+    }
+}");
         }
 
         [Fact]
@@ -141,7 +193,7 @@ GetCA1307CSharpResultsAt(12, 20, $"string.CompareTo({ObjectArgType})",
         }
 
         [Fact]
-        public async Task CA1307_OverloadTests_CSharp()
+        public async Task CA1307_OverloadTests_StringFirstParam_CSharp()
         {
             await VerifyCS.VerifyAnalyzerAsync(@"
 using System;
@@ -170,6 +222,34 @@ public class StringComparisonTests
 GetCA1307CSharpResultsAt(9, 9, "StringComparisonTests.DoNothing(string)",
                                "StringComparisonTests.NonString()",
                                "StringComparisonTests.DoNothing<T>(string, System.StringComparison)"));
+        }
+
+        [Theory, WorkItem(2581, "https://github.com/dotnet/roslyn-analyzers/issues/2581")]
+        [InlineData("char")]
+        [InlineData("int")]
+        [InlineData("object")]
+        [InlineData("StringComparisonTests")]
+        public async Task CA1307_OverloadTests_NonStringFirstParam_CSharp(string firstParamType)
+        {
+            await VerifyCS.VerifyAnalyzerAsync($@"
+using System;
+using System.Globalization;
+
+public class StringComparisonTests
+{{
+    public void NonString({firstParamType} p)
+    {{
+        [|DoNothing(p)|];
+    }}
+
+    public void DoNothing({firstParamType} p)
+    {{
+    }}
+
+    public void DoNothing({firstParamType} p, StringComparison strCompare)
+    {{
+    }}
+}}");
         }
 
         [Fact]

--- a/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SpecifyStringComparisonTests.cs
+++ b/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/SpecifyStringComparisonTests.cs
@@ -80,6 +80,7 @@ GetCA1307CSharpResultsAt(12, 16, "string.StartsWith(string)",
                                  "string.StartsWith(string, System.StringComparison)"));
         }
 
+#if NETCOREAPP // EndsWith(char) and StartsWith(char) overloads don't exist in .NET Framework 4.7.2
         [Fact, WorkItem(2581, "https://github.com/dotnet/roslyn-analyzers/issues/2581")]
         public async Task CA1307_StringWithCharTests_CSharp()
         {
@@ -96,6 +97,7 @@ public class StringComparisonTests
     }
 }");
         }
+#endif
 
         [Fact]
         public async Task CA1307_StringIndexOfStringTests_CSharp()

